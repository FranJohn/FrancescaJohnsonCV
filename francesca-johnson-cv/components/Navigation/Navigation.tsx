'use client'
/**
 * React functional component representing a navigation bar with dropdown functionality.
 *
 * @component
 * @example
 * // Example usage of Navigation component
 * <Navigation title="My Navigation" pages={['Home', 'About', 'Services']} />
 *
 * @param {Object} props - The properties passed to the component.
 * @param {string} props.title - The title of the navigation bar.
 * @param {Array} props.pages - An array of strings representing the pages in the navigation bar.
 * @returns {JSX.Element} Rendered Navigation component.
 */
import React, { useState } from 'react';
import styles from './Navigation.module.css';

/**
 * @typedef {Object} NavigationProps - The props for the Navigation component.
 * @property {string} title - The title of the navigation bar.
 * @property {string[]} pages - An array of strings representing the pages in the navigation bar.
 */
interface NavigationProps {
    title: string;
    pages: string[];
}


/**
 * Navigation component functional definition.
 *
 * @param {NavigationProps} props - The properties passed to the component.
 * @returns {JSX.Element} Rendered Navigation component.
 */
const Navigation: React.FC<NavigationProps> = ({ title, pages }) => {
    const [isOpen, setIsOpen] = useState(false);

    /**
    * Toggles the dropdown visibility when the hamburger icon is clicked.
    */
    const toggleDropdown = () => {
        setIsOpen(!isOpen);
        console.log("Toggle dropdown");
    };

    /**
     * Maps the pages to dropdown items.
     *
     * @type {Array<JSX.Element>} Array of JSX elements representing dropdown items.
     */ 
    const dropdownItems = pages.map((item, index) => (
        <div
            key={index}
            className={styles.dropdown__item}
            onClick={() => toggleDropdown()}
        >
            {item}
        </div>
    ));

    return (
    <>
        <div className={`${styles.nav__container} ${isOpen ? styles.open : ''}`}>
            <nav className={`${styles.navigation} ${isOpen ? styles.open : ''}`}>
                <div className={styles.logo}>{title}</div>
                <div className={styles.links}>
                    {pages.map((page, index) => (
                        <a key={index} href={`/${page.toLowerCase().replace(' ', '-')}`} data-testid={`nav-link-${page.toLowerCase()}`}>
                            {page}
                        </a>
                    ))}
                </div>
            </nav>
            <div className={styles.dropdown__container} >
                <div className={`${styles.dropdown} ${isOpen ? styles.active : ''}`}>
                    <div className={`${styles.dropdown__hamburger}`} onClick={toggleDropdown} data-testid="hamburger">
                        <div className={`${styles['hamburger-line']}`}></div>
                        <div className={`${styles['hamburger-line']}`}></div>
                        <div className={`${styles['hamburger-line']}`}></div>
                    </div>
                </div>          
            </div>
        </div>
        <div className={`${styles.dropdown} ${isOpen ? styles.active : ''}`} data-testid="dropdown" >
                <div className={`${styles.dropdown__items} `} >
                    {dropdownItems}
                </div>
        </div>  
    </>
  );
};

export default Navigation;
