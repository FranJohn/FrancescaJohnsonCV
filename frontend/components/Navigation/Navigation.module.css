.head__container {
    border-bottom: #DA7b93 1px solid;
    display:block;
    width:100%;
}

.nav__container {
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    padding: 2rem;
    padding-bottom: 1rem;
}

.navigation {
    display: flex;
    flex-direction: row;
    padding: 0rem;
    background-color: rgba(28, 51, 52, 0);
    color: white;
    position: relative;
    width: auto;
    flex:1
}

.logo {
    font-size: 1.5rem;
    background-color: rgba(28, 51, 52, 0);
    transition: background-color 1s ease;
    text-align: center;
    width: 100%;
}

.links {
    display: flex;
    gap: 2rem;
}

.dropdown_hamburger {
    display: none;
    flex-direction: column;
    cursor: pointer;
    position: relative;
}

.dropdown__container {
    display: none;
  }

.dropdown {
     display: none;
}

.page:hover,
.dropdown__item:hover,
.page.current,
.dropdown__item.current {
  font-weight: bold;
}


@media (max-width: 768px) {
    .nav__container {
        padding-bottom: 0rem;
    }
    .nav__container.open {
        background-color: rgba(28, 51, 52, 0.8);
    }

    .links {
        display: none;
    }


    .dropdown__container {
        position: relative;
        margin-left: auto;
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        cursor: pointer;
        width: fit-content;
    }

    .dropdown {
        position: relative;
        display: inline-block;
        cursor: pointer;
        /* margin-left: auto; */
        align-items: center;
        width:100%;
        text-align: center;
    }
  
    /* Hamburger styles */
    .dropdown__hamburger {
        width: 30px;
        height: 20px;
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        cursor: pointer;
    }
  
    .hamburger-line {
        width: 100%;
        height: 2px;
        background-color: #ffffff;
        transition: 0.3s;
    }
  
    /* Open state for hamburger */
    .hamburger-line.open {
        transform: rotate(45deg);
        background-color: #fff;
    }
  
    /* Dropdown items */
    .dropdown__items {
        background-color: rgba(28, 51, 52, 0.8);
        position: relative;
        visibility: hidden;
        max-height: 0;
        top:100%;
        width:100%;
        transition: max-height 0.5s ease-in-out, visibility 0.5s ease-in-out;
        overflow: hidden;
        cursor: pointer;
        padding: 10px 18px;
        opacity: 0;
        transition: opacity 0.5s ease-in-out;
        align-items: center;
        justify-content: center;
        align-content: center;
    }
  
    /* Active state for dropdown */
    .dropdown.active .dropdown__items {
        right: 0px;
        visibility: visible;
        max-height: 290px;
        transition: max-height 0.5s ease-in-out, visibility 0.5s ease-in-out;
        opacity: 1;
        transition-delay: 0.1s;
        align-items: center;
        justify-content: center;
        align-content: center;
    } 

    .dropdown__item {
        width: 100%;
        display: block;
    }

    .dropdown__item:hover {
        font-weight: bold; 
    }
}