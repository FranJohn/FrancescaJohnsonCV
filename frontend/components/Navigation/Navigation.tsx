'use client'
/**
 * React functional component representing a navigation bar with dropdown functionality.
 *
 * @component
 * @example
 * // Example usage of Navigation component
 * <Navigation title="My Navigation" pages={['Home', 'About', 'Services']} />
 *
 * @param {Object} props - The properties passed to the component.
 * @param {string} props.title - The title of the navigation bar.
 * @param {Array} props.pages - An array of strings representing the pages in the navigation bar.
 * @returns {JSX.Element} Rendered Navigation component.
 */
import React, { useState, useEffect } from 'react';
import styles from './Navigation.module.css';

/**
 * @typedef {Object} NavigationProps - The props for the Navigation component.
 * @property {string} title - The title of the navigation bar.
 * @property {string[]} pages - An array of strings representing the pages in the navigation bar.
 */
interface NavigationProps {
    title: string;
    pages: string[];
}


/**
 * Navigation component functional definition.
 *
 * @param {NavigationProps} props - The properties passed to the component.
 * @returns {JSX.Element} Rendered Navigation component.
 */
const Navigation: React.FC<NavigationProps> = ({ title, pages }) => {
    const [isOpen, setIsOpen] = useState(false);
    const [currentRoute, setCurrentRoute] = useState('home');

    const set_route = (page: string) => {
        // Set the current route when the route changes
        setCurrentRoute(page);
        toggleDropdown();
      };

    /**
    * Toggles the dropdown visibility when the hamburger icon is clicked.
    */
    const toggleDropdown = () => {
        setIsOpen(!isOpen);
        console.log("Toggle dropdown");
    };

    /**
     * Maps the pages to dropdown items.
     *
     * @type {Array<JSX.Element>} Array of JSX elements representing dropdown items.
     */ 
    const dropdownItems = pages.map((item, index) => (
        <a
            key={index}
            href={`/${item.toLowerCase().replace(' ', '-')}`}
            className={`${styles.dropdown__item} ${currentRoute === `/${item.toLowerCase().replace(' ', '-')}` ? styles.current : ''}`}
            onClick={() => set_route(item)}
        >
            {item}
        </a>
    ));

    return (
    <div className={styles.head__container}>
        <div className={`${styles.nav__container} ${isOpen ? styles.open : ''}`}>
            <nav className={`${styles.navigation} ${isOpen ? styles.open : ''}`}>
                <div className={styles.logo}>{title}</div>
                <div className={styles.links}>
                    {pages.map((page, index) => (
                        <a
                            key={index}
                            className={`${styles.page} ${currentRoute === `/${page.toLowerCase().replace(' ', '-')}` ? styles.current : ''}`}
                            href={`/${page.toLowerCase().replace(' ', '-')}`}
                            onClick={() => set_route(page)}
                            data-testid={`nav-link-${page.toLowerCase()}`}
                        >                            
                        {page}
                        </a>
                    ))}
                </div>
            </nav>
            <div className={styles.dropdown__container} >
                <div className={`${styles.dropdown} ${isOpen ? styles.active : ''}`}>
                    <div className={`${styles.dropdown__hamburger}`} onClick={toggleDropdown} data-testid="hamburger">
                        <div className={`${styles['hamburger-line']}`}></div>
                        <div className={`${styles['hamburger-line']}`}></div>
                        <div className={`${styles['hamburger-line']}`}></div>
                    </div>
                </div>          
            </div>
        </div>
        <div className={`${styles.dropdown} ${isOpen ? styles.active : ''}`} data-testid="dropdown" >
                <div className={`${styles.dropdown__items} `} >
                    {dropdownItems}
                </div>
        </div>  
    </div>
  );
};

export default Navigation;
