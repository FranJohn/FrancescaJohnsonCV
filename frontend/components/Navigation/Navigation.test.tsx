/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom';
import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import Navigation from './Navigation';

/**
 * Test suite for the Navigation component.
 */
describe('Navigation Component', () => {
  const title = 'Test Navigation';
  const pages = ['Home', 'About', 'Services'];

  /**
   * Test case to check if the navigation component renders
   * with correct title, links, and dropdown items.
   */
  test('renders navigation component with correct title, links, and dropdown items', async () => {
    const { getByTestId, getAllByText } = render(<Navigation title={title} pages={pages} />);
    
    await waitFor(() => {
      pages.forEach((page) => {
        // Check navigation links
        const navLinks = getAllByText(page);
        navLinks.forEach((link) => {
          expect(link).toBeInTheDocument();
          expect(getByTestId(`nav-link-${page.toLowerCase().replace(' ', '-')}`)).toBeInTheDocument();
        });

        // Check dropdown items
        const dropdownItems = getAllByText(page);
        dropdownItems.forEach((item) => {
          expect(item).toBeInTheDocument();
        });
      });
    });
  });

  /**
   * Test case to check if the dropdown is closed by default.
   */
  test('dropdown is closed by default', () => {
    const { queryByTestId } = render(<Navigation title={title} pages={pages} />);
    expect(queryByTestId('hamburger')).toBeInTheDocument();
    const dropdown = queryByTestId('dropdown');
    expect(dropdown).not.toHaveClass('active');
  });

  /**
   * Test case to check if the dropdown opens and closes
   * on hamburger click.
   */
  test('dropdown opens and closes on hamburger click', () => {
    const { getByTestId } = render(<Navigation title={title} pages={pages} />);
    const hamburger = getByTestId('hamburger');
    const dropdown = getByTestId('dropdown');

    // Initially, the dropdown should be closed
    expect(dropdown).not.toHaveClass('active');

    // Click on the hamburger to open the dropdown
    fireEvent.click(hamburger);
    expect(dropdown).toHaveClass('active');

    // Click on the hamburger again to close the dropdown
    fireEvent.click(hamburger);
    expect(dropdown).not.toHaveClass('active');
  });
});
