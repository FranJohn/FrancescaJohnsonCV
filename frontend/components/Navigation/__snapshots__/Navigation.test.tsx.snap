// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Component matches snapshot 1`] = `
<div>
  <nav
    class="navigation"
  >
    <div
      class="logo"
    >
      Your CV
    </div>
    <div
      class="links"
    >
      <a
        data-testid="nav-link-home"
        href="/home"
      >
        Home
      </a>
      <a
        data-testid="nav-link-about"
        href="/about"
      >
        About
      </a>
      <a
        data-testid="nav-link-services"
        href="/services"
      >
        Services
      </a>
    </div>
    <div
      class="hamburger "
      data-testid="hamburger-menu"
    >
      <div
        class="hamburger-line"
      />
      <div
        class="hamburger-line"
      />
      <div
        class="hamburger-line"
      />
    </div>
    <div
      class="dropdown "
      data-testid="dropdown-menu"
    >
      <a
        data-testid="dropdown-link-home"
        href="/home"
      >
        Home
      </a>
      <a
        data-testid="dropdown-link-about"
        href="/about"
      >
        About
      </a>
      <a
        data-testid="dropdown-link-services"
        href="/services"
      >
        Services
      </a>
    </div>
  </nav>
</div>
`;
